src/__tests__/logql-query.spec.ts

```diff
@@ -3,59 +3,341 @@ import { LogQLQuery } from '../logql-query';
+        selector: { label: 'foo', operator: '=', value: '"var"' },
+        expected: '{ baz=~"zad" }',
+      },
+    ].forEach(({ query, selector, expected }) => {
+      const logql = new LogQLQuery(query);
+      expect(logql.removeMatcher(selector).toString()).toEqual(expected);
+    });
+  });
+
+  it('should append a new pipeline stage', () => {
+    [
+      {
+        query: '{ foo="var" }',
+        pipeline: { operator: '|', value: 'json' },
+        expected: '{ foo="var" } | json',
+      },
+      {
+        query: '{ foo="var" }| json',
+        pipeline: { operator: '|', value: 'level="err|error"' },
+        expected: '{ foo="var" } | json | level="err|error"',
+      },
+      {
+        query: '{ foo="var" }| json | level="err|error"',
+        pipeline: { operator: '|', value: 'level="err|error"' },
+        expected: '{ foo="var" } | json | level="err|error"',
+      },
+    ].forEach(({ query, pipeline, expected }) => {
+      const logql = new LogQLQuery(query);
+      expect(
+        logql.addPipelineStage(pipeline, { placement: 'end' }).toString(),
+      ).toEqual(expected);
+    });
+  });
+
+  it('should prepend a pipeline stage', () => {
+    [
+      {
+        query: '{ foo="var" }',
+        pipeline: { operator: '|', value: 'json' },
+        expected: '{ foo="var" } | json',
       },
       {
-        query: '{ foo=var }',
-        pipeline: ['json', 'level="err|error"'],
-        expected: '{ foo=var } | json | level="err|error"',
+        query: '{ foo="var" }| json',
+        pipeline: { operator: '|', value: 'level="err|error"' },
+        expected: '{ foo="var" } | level="err|error" | json',
+      },
+      {
+        query: '{ foo="var" } | json',
+        pipeline: [
+          { operator: '|=', value: '"match this text"' },
+          { operator: '|', value: 'level="err|error"' },
+        ],
+        expected:
+          '{ foo="var" } |= "match this text" | level="err|error" | json',
       },
     ].forEach(({ query, pipeline, expected }) => {
       const logql = new LogQLQuery(query);
-      expect(logql.appendPipeline(pipeline).toString()).toEqual(expected);
+      expect(
+        logql.addPipelineStage(pipeline, { placement: 'start' }).toString(),
+      ).toEqual(expected);
+    });
+  });
+
+  it('should remove a pipeline stage', () => {
+    [
+      {
+        query: '{ foo="var" }',
+        pipeline: { operator: '|' },
+        expected: '{ foo="var" }',
+      },
+      {
+        query: '{ foo="var" } | json | level="err|error"',
+        pipeline: { operator: '|' },
+        expected: '{ foo="var" }',
+      },
+      {
+        query: '{ foo="var" } | json | level="err|error"',
+        pipeline: { value: 'json' },
+        expected: '{ foo="var" } | level="err|error"',
+      },
+      {
+        query: '{ foo="var" } | json | level="err|error"',
+        pipeline: {},
+        matchOptions: { matchLabel: 'level' },
+        expected: '{ foo="var" } | json',
+      },
```

Comment:
Could we also add a case to test what happens when the `matchLabel` doesn't match an existing label?